dev-tools/mage/dockerbuilder.go

```diff
@@ -0,0 +1,143 @@
+// under the License.
+
+package mage
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/magefile/mage/sh"
+
+	"github.com/pkg/errors"
+)
+
+type dockerBuilder struct {
+	PackageSpec
+
+	buildDir string
+	beatDir  string
+}
+
+func newDockerBuilder(spec PackageSpec) (*dockerBuilder, error) {
+	buildDir := filepath.Join(spec.packageDir, "docker-build")
+	beatDir := filepath.Join(buildDir, "beat")
+
+	return &dockerBuilder{
+		PackageSpec: spec,
+		buildDir:    buildDir,
+		beatDir:     beatDir,
+	}, nil
+}
+
+func (b *dockerBuilder) Build() error {
+	if err := os.RemoveAll(b.buildDir); err != nil {
+		return errors.Wrapf(err, "failed to clean existing build directory %s", b.buildDir)
+	}
+
+	if err := b.copyFiles(); err != nil {
+		return err
+	}
+
+	if err := b.prepareBuild(); err != nil {
+		return errors.Wrap(err, "failed to prepare build")
+	}
+
+	tag, err := b.dockerBuild()
+	if err != nil {
+		return errors.Wrap(err, "failed to build docker")
+	}
+
+	if err := b.dockerSave(tag); err != nil {
+		return errors.Wrap(err, "failed to save docker as artifact")
+	}
+
+	return nil
+}
+
+func (b *dockerBuilder) modulesDirs() []string {
+	var modulesd []string
+	for _, f := range b.Files {
+		if f.Modules {
+			modulesd = append(modulesd, f.Target)
+		}
+	}
+	return modulesd
+}
+
+func (b *dockerBuilder) copyFiles() error {
+	for _, f := range b.Files {
+		target := filepath.Join(b.beatDir, f.Target)
+		if err := Copy(f.Source, target); err != nil {
+			return errors.Wrapf(err, "failed to copy from %s to %s", f.Source, target)
+		}
+	}
+	return nil
+}
+
+func (b *dockerBuilder) prepareBuild() error {
+	elasticBeatsDir, err := ElasticBeatsDir()
+	if err != nil {
+		return err
+	}
+	templatesDir := filepath.Join(elasticBeatsDir, "dev-tools/packaging/templates/docker")
+
+	data := map[string]interface{}{
+		"ModulesDirs": b.modulesDirs(),
+	}
+
+	return filepath.Walk(templatesDir, func(path string, info os.FileInfo, _ error) error {
+		if !info.IsDir() {
+			target := strings.TrimSuffix(
+				filepath.Join(b.buildDir, filepath.Base(path)),
+				".tmpl",
+			)
+
+			err = b.ExpandFile(path, target, data)
+			if err != nil {
+				return errors.Wrapf(err, "expanding template '%s' to '%s'", path, target)
+			}
+		}
+		return nil
+	})
+}
+
+func (b *dockerBuilder) dockerBuild() (string, error) {
+	tag := fmt.Sprintf("%s:%s", b.Name, b.Version)
+	if repository, _ := b.ExtraVars["repository"]; repository != "" {
+		tag = fmt.Sprintf("%s/%s", repository, tag)
+	}
+	return tag, sh.Run("docker", "build", "-t", tag, b.buildDir)
+}
+
+func (b *dockerBuilder) dockerSave(tag string) error {
+	// Save the container as artifact
+	outputFile := b.OutputFile
+	if outputFile == "" {
+		outputTar, err := b.Expand(defaultBinaryName + ".docker.tar")
+		if err != nil {
+			return err
+		}
+		outputFile = filepath.Join(distributionsDir, outputTar)
+	}
+	if err := sh.Run("docker", "save", "-o", outputFile, tag); err != nil {
```

Comment:
I like this but it doesn't appear to be required for releases - the publisher currently does this + compression

Reply:
Oh, I though we also needed to generate the artifact and publisher does `docker save` and `docker push`.
I am using it also for sanity checks, so I'd leave it if it is not a problem.

Reply:
Not a problem from my perspective, agreed on keeping it.